src/services/firm_quote_validator.ts

```diff
@@ -0,0 +1,147 @@
+import { assetDataUtils, ERC20AssetData } from '@0x/order-utils';
+import { In } from 'typeorm';
+import { Repository } from 'typeorm/repository/Repository';
+
+import { ONE_SECOND_MS } from '../constants';
+import { MakerBalanceChainCacheEntity } from '../entities/MakerBalanceChainCacheEntity';
+import { logger } from '../logger';
+
+
+const BIG_NUMBER_ZERO = new BigNumber(0);
+const THRESHOLD_CACHE_EXPIRED_MS = 2 * 60 * ONE_SECOND_MS;
+
+
+export class PostgresBackedFirmQuoteValidator implements RfqtFirmQuoteValidator {
+    private readonly _chainCacheRepository: Repository<MakerBalanceChainCacheEntity>;
+    private readonly _cacheExpiryThresholdMs: number;
+
+    private static _calculateTakerFillableAmountsFromQuotes(quotes: SignedOrder[], makerLookup: {[key: string]: BigNumber}): {makerAddressesToAddToCache: string[]; takerFillableAmounts: BigNumber[]} {
+        const makerAddressesToAddToCacheSet: Set<string> = new Set();
+        const takerFillableAmounts =  quotes.map(quote => {
+            const makerTokenBalanceForMaker: BigNumber | undefined = makerLookup[quote.makerAddress];
+
+            // TODO: Add Prometheus hooks
+            if (makerTokenBalanceForMaker === undefined) {
+                makerAddressesToAddToCacheSet.add(quote.makerAddress);
+                return quote.takerAssetAmount;
+            }
+
+            // Order is fully fillable, because Maker has 100% of the assets
+            if (makerTokenBalanceForMaker.gte(quote.makerAssetAmount)) {
+                return quote.takerAssetAmount;
+            }
+
+            // Order is empty, return zero
+            if (quote.makerAssetAmount.lte(0)) {
+                return BIG_NUMBER_ZERO;
+            }
+
+            // Order is partially fillable, because Maker has a fraction of the assets
+            const partialFillableAmount = makerTokenBalanceForMaker.times(quote.takerAssetAmount).div(quote.makerAssetAmount).integerValue(BigNumber.ROUND_DOWN);
+            if (!partialFillableAmount.isFinite()) {
+                logger.error(`Calculated maker token balance is infinite, which caused the partialFillableAmount to be infinite. This should never happen`);
+                return BIG_NUMBER_ZERO;
+            }
+            return partialFillableAmount;
+        });
+        const makerAddressesToAddToCache = Array.from(makerAddressesToAddToCacheSet);
+        return {
+            makerAddressesToAddToCache,
+            takerFillableAmounts,
+        };
+    }
+
+    constructor(chainCacheRepository: Repository<MakerBalanceChainCacheEntity>, cacheExpiryThresholdMs: number = THRESHOLD_CACHE_EXPIRED_MS) {
+        this._chainCacheRepository = chainCacheRepository;
+        this._cacheExpiryThresholdMs = cacheExpiryThresholdMs;
+    }
+
+    // tslint:disable-next-line: prefer-function-over-method
+    public async getRFQTTakerFillableAmounts(quotes: SignedOrder[]): Promise<BigNumber[]> {
```

Comment:
sometimes we use `RFQT` in all caps and sometimes we use `Rfqt`

Reply:
Also, don't we usually add `Async` as a suffix to async methods? maybe the linter doesnt catch that anymore..